oak_crypto/src/verifier.rs

```diff
@@ -0,0 +1,32 @@
+//
+// Copyright 2023 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use anyhow::Context;
+
+pub trait Verifier {
+    fn verify(&self, message: &[u8], signature: &crate::signer::Signature) -> anyhow::Result<()>;
+}
+
+impl Verifier for p256::ecdsa::VerifyingKey {
+    fn verify(&self, message: &[u8], signature: &crate::signer::Signature) -> anyhow::Result<()> {
+        let parsed_signature = p256::ecdsa::Signature::from_slice(&signature.signature)
+            .map_err(anyhow::Error::msg)
+            .context("could not parse signature")?;
+        <p256::ecdsa::VerifyingKey as p256::ecdsa::signature::Verifier<
+            p256::ecdsa::Signature,
+        >>::verify(self, message, &parsed_signature).map_err(anyhow::Error::msg)
```

Comment:
Are we sure we want to implement our own `verify` method on top of an existing type that also has its own `verify` method? Maybe we should reconsider the design and just use our own opaque types as the interface (even if we implement them in terms of p256 or other crates, but without exposing them).

Reply:
I think this is idiomatic way to go! 

The important part is the trait definition, which uses our own types only:

```
pub trait Verifier {
    fn verify(&self, message: &[u8], signature: &crate::signer::Signature) -> anyhow::Result<()>;
}
```

Multiple structs may implement this trait, the implementation provided for `p256::ecdsa::VerifyingKey` is just one. It just means that you would be able use this type for oak_crypto verification directly, without first having to wrap it in another struct.

In the SDK, we of course wouldn't expose the `p256::ecdsa::VerifyingKey`. Instead the type signature could look eg like this:

```
pub struct InstanceSigner {
    key: &'static SigningKey,
}

impl InstanceSigner {
    pub fn create() -> anyhow::Result<Self> {
        ...
    }
}

impl Signer for InstanceSigner {
    fn sign(&self, message: &[u8]) -> anyhow::Result<oak_crypto::signer::Signature> {
        ...
    }
}

impl Verifier for InstanceSigner {
    fn verify(
        &self,
        message: &[u8],
        signature: &oak_crypto::signer::Signature,
    ) -> anyhow::Result<()> {
        let verifying_key = p256::ecdsa::VerifyingKey::from(self.key);
        <p256::ecdsa::VerifyingKey as oak_crypto::verifier::Verifier>::verify(
            &verifying_key,
            message,
            signature,
        )
    }
}
```

I understand that in C++, this would look different. Instead of a trait, oak crypto would expose a class. But traits are more powerful & versatile. It doesn't come with the tradeoffs of needing to deal with a series of wrapper classes.

Reply:
If we really want to also export our own struct that implements this trait, and this trait only, we could. But I think that would just create overhead in places where it's used. 


eg in oak_crypto/src/verifier.rs we could add the following:


```
struct VerifierKeyHandle {
    inner: p256::ecdsa::VerifyingKey,
}

impl Verifier for VerifierKeyHandle {
    fn verify(&self, message: &[u8], signature: &crate::signer::Signature) -> anyhow::Result<()> {
        self.inner.verify(message, signature)
    }
}
```

Reply:
> > `impl Verifier for InstanceSigner`
> 
> Why do we want signer to verify? The verifier is only needed for the client (or anyone who receives a public key from protobuf)

the example was for illustrative purposes, not a proposal for an actual change to the SDK. http://b/314749025 mentioned we may want to support verifying as part of the SDK.

Reply:
@tiziano88 may I proceed with merging the PR as is? We can revisit this later.

Reply:
definitely, we can discuss separately (if useful)